### DISCUSSION TITLE
[RFC] is there a way for turborepo to detect if sub-processes are also invoking turbo and protect against infinite process spawning cycles?

### DISCUSSION THREAD

**Discussion by NullVoxPopuli** [vercel/turborepo#7467] [Ideas]
### What version of Turborepo are you using?

1.9.3

### What package manager are you using / does the bug impact?

pnpm

### What operating system are you using?

Linux

### Describe the Bug

When running a few commands, it seems both RAM and CPU hit limits.
![image](https://user-images.githubusercontent.com/199018/236685506-a3f3dc4a-b9be-445a-b06b-d37ef9d28971.png)

The system freezes for a time, and then once it recovers, the terminal window that ran the turbo commands is killed.

### Expected Behavior

No memory leak.

### To Reproduce

- `git clone git@github.com:NullVoxPopuli/glimdown.git`
- `cd glimdown`
- `git checkout implementation`
- `pnpm lint:fix` maybe 4-5 times?

### Reproduction Repo

https://github.com/NullVoxPopuli/glimdown

### Unintended side-effects

This is still being explored, and it seems to be a bit random -- just.. "whatever is happening" at the time of the process getting killed by the OS

But so far I've ran in to two things:
 - turbo daemon pidfile remains, and there is a warning printed about the pid file, but no detected running daemon
```
 WARNING  failed to contact turbod. Continuing in standalone mode: connection to turbo daemon process failed. Please ensure the following:
	- the process identified by the pid in the file at /tmp/turbod/86cc8c486dcd6ee2/turbod.pid is not running, and remove /tmp/turbod/86cc8c486dcd6ee2/turbod.pid
	- check the logs at /home/nvp/.local/share/turborepo/logs/86cc8c486dcd6ee2-glimdown.log
	- the unix domain socket at /tmp/turbod/86cc8c486dcd6ee2/turbod.sock has been removed
	You can also run without the daemon process by passing --no-daemon
```
 - git has a stray lockfile 
 ```
 ❯ git add .
fatal: Unable to create '/home/nvp/Development/NullVoxPopuli/glimdown/.git/index.lock': File exists.

Another git process seems to be running in this repository, e.g.
an editor opened by 'git commit'. Please make sure all processes
are terminated then try again. If it still fails, a git process
may have crashed in this repository earlier:
remove the file manually to continue.
```

**Comment by NullVoxPopuli** [comment]
![image](https://user-images.githubusercontent.com/199018/236686248-f51fb7b7-62fc-4d9f-8b09-7e169e59be44.png)


I was watching more closely and had `top` running next to the activity monitor.

I wasn't able to take a screenshot of what I saw as the machine hung, but but it seems `go-turbo` is involved, as well as _lots_ of `node` processes.

I know that each node process is, by default, allowed to consume 4GB of RAM. 

Since it seems that turbo is spawning a node process for each of the 24 available threads, that's 96 required GB of RAM... assuming that node is wanting to use all of the 4GB of ram per process.

I'm only running `eslint` and `prettier` here, so this  feels a bit goofy. I have very small projects ( less than 10 files each ).

**Comment by NullVoxPopuli** [comment]
Limiting the concurrency didn't help either, I had a spike across all CPUs still:
```
❯ pnpm lint:fix --concurrency 8

> glimdown-monorepo-root@0.0.0 lint:fix /home/nvp/Development/NullVoxPopuli/glimdown
> turbo run lint:fix --output-logs errors-only "--concurrency" "8"
```
![image](https://user-images.githubusercontent.com/199018/236686737-c553b79c-1264-41a0-9f86-7a1b16fac50c.png)

**Comment by NullVoxPopuli** [comment]
when this happens with concurrency 8, I feel like `turbo` is having a problem resolving the dependency graph, and has gotten stuck in a cycle or something.
This is the only config I can think of from my turbo.json that could be related:
```
    "lint:fix": {
      "outputs": [],
      "dependsOn": ["lint:js:fix", "lint:hbs:fix", "lint:prettier:fix"]
    },
    "lint:js:fix": {
      "outputs": ["lint:prettier:fix"] // I realize now that this wasn't doing anything.
      // I've since switched this to "dependsOn", because eslint and prettier could 
      // conflict with eachother if running in prallel.
    },
    "lint:hbs:fix": {
      "outputs": []
    },
    "lint:prettier:fix": {
      "outputs": []
    }
```

gonna remove `lint:prettier:fix` from `lint:fix` to see if that's the issue -- even though this isn't a cycle.
No change. hmmm

While my system still hung a bit due to running out of RAM, CPU usage was more desirable.
![image](https://user-images.githubusercontent.com/199018/236687229-d7f05857-4cfe-4230-bb7f-e48645fd12b6.png)

**Comment by NullVoxPopuli** [comment]
Ok, so it looks like I had a cycle.

A few of my packages defined `lint:fix`, which is also used for `turbo`.

Here is how I resolved: https://github.com/NullVoxPopuli/glimdown/pull/4/files#diff-f8de965273949793edc0fbfe249bb458c0becde39b2e141db087bcbf5d4ad5e3R1
Both memory and cpu usage is now _much_ more reasonable, and matches the number of relevant projects.

So.. _now I'm more confused_ -- how this isn't an "all the time problem".

(previously:)
I run 
- `pnpm lint:fix` at the monorepo root
- aliased in package.json: `"lint:fix": "turbo run lint:fix --output-logs errors-only",`
- my `turbo.json` defines `lint:fix` as:
    ```
    "lint:fix": {
      "outputs": [],
      "dependsOn": ["lint:js:fix", "lint:hbs:fix"]
    },
    "lint:js:fix": {
      "outputs": [],
      "dependsOn": ["lint:prettier:fix"]
    },
    "lint:hbs:fix": {
      "outputs": []
    },
    "lint:prettier:fix": {
      "outputs": []
    }
    ```
- so I would expect `lint:fix` to launch `lint:js:fix`, `lint:prettier:fix`, and `lint:hbs:fix`
- but what I did not expect, and I'm not sure why, tbh, is that if a package defines `lint:fix`, that is also ran.
- many of my packages had this: `    "lint:fix": "pnpm -w exec lint fix",`
- and that is an alias to a local dev-package which provides the `lint` bin (because proper lint configs are complicated, and I like abstracting it out of each package, so that I can manage everything in one place and have the changes easily propagated everywhere at once)
    the script:
    ```js
    import path from "node:path";

    import { packageJson, project } from "ember-apply";
    import { execaCommand } from "execa";

    const [, , command] = process.argv;
    // process.cwd() is whatever pnpm decides to do
    //
    // Other options:
    //  PNPM_SCRIPT_SRC_DIR
    //  OLDPWD
    const cwd = process.env["INIT_CWD"];

    const root = await project.gitRoot();
    const manifest = await packageJson.read(cwd);
    const relative = path.relative(root, cwd);

    console.debug(`${manifest.name} :: within ${relative}`);

    async function run() {
      switch (command) {
        case "prettier:fix":
          return execaCommand(
            `pnpm prettier -w . --cache --cache-strategy content`,
            { cwd }
          );
        case "prettier":
          return execaCommand(`pnpm prettier -c .`, { cwd });
        case "js:fix":
          return execaCommand(
            `pnpm eslint . --fix --cache --cache-strategy content`,
            { cwd }
          );
        case "js":
          return execaCommand(`pnpm eslint .`, { cwd });
        case "fix":
          return execaCommand(`pnpm turbo lint:fix`);
        default:
          return execaCommand(`pnpm turbo lint`);
      }
    }

    await run();
  
    ```
- so `lint fix` runs `pnpm turbo lint:fix` ... and we have a cycle.



## The real problem / question

Is there a way for turbo to detect if this nested-process type situation is going to cause a cycle?
maybe by allowing only one process at a time? or maybe having the multiple processes communicate their task-graph intent so that they can collectively resolve cycles?

**Comment by chris-olszewski** [comment]
This is a hard thing to get correct. We [currently check scripts](https://github.com/vercel/turbo/pull/4325/files#diff-0d8c9dc4ff5f7b55c5650a08fe73a82a8813bcc293fbfdd08992046ab47ee6daL241) and try to warn if a user is invoking turbo as part of a turbo task. As you've discovered this isn't fool proof as an additional layer of scripts can easily bypass this check. 

There might be some way to "monitor" process that watches the process tree and warns if it sees a turbo process in the task graph, but this would require quite the overhaul to our engine and I don't expect for us to prioritize it. Sorry for the poor experience you had.
